### ISSUE TITLE
Gitlab-based sync process on pre-existing workspace repo with submodules intermittently throws "Skipping object checkout, Git LFS is not installed."

### ISSUE THREAD

**Issue by niasharper** [git-lfs/git-lfs#3237]
## Overview

Our gitlab infrastructure uses a fetch strategy instead of a burn and clone strategy on the workspace.  Gitlab internally has LFS support and submodule support enabled.  With tens of builds per node per day going across our 15-runners build infrastructure, we have been seeing intermittent errors from git lfs spring up approximately once every week or two.

Most all of our repos are leveraging submodules.  The parent repo is typically not using LFS.  The submodule is LFS-Enabled.

## Problem/Error Received

When a build node goes into a bad state, the following is consistently seen across sync'd local repos and projects until LFS is re-initialized on the problem node:
```
(Gitlab fetches changes)
Fetching changes...

(Gitlab checks out tree state)
Checking out <SHA> as <Branch Ref>...

(Gitlab updates/synch)
Updating/initializing submodules recursively...

(.gitlab-ci.yml script block)
$ git submodule foreach --recursive git clean -fdx
Entering 'repo/path/to/submodule'
$ git submodule update --init --recursive
$ git lfs pull
Skipping object checkout, Git LFS is not installed.
```

## Work-Around Solution

Corrective Work-Around Action:
- Log into each problem build node
- Re-run git lfs install:
```
$ git lfs install
Error: Failed to call git rev-parse --git-dir --show-toplevel: "fatal: not a git repository (or any of the parent directories): .git\n"
Git LFS initialized.
```

Future gitlab builds on this node will then succeed and sync correctly until the problem occurs again and the "LFS is not installed" error is seen in the logs.

## Expected behavior

Git LFS should not de-initialize on pre-existing infrastructure.  Git LFS should be stable, reliable, and predictable.

Git LFS should be able to work with typical submodule workflows

## Build Infrastructure Details

Builds occur on HIgh Sierra build nodes with a concurrency of 2 parallel builds per build node under the same active logged in user.

#### OS X
```
      System Version: macOS 10.13.4 (17E199)
      Kernel Version: Darwin 17.5.0
```
#### Git
```
$ git --version
git version 2.18.0
$ git-lfs --version
git-lfs/2.4.2 (GitHub; darwin amd64; go 1.10.2; git 6f4b2e98) 
```

#### Gitlab Runner Version
```
$ gitlab-runner-darwin-amd64 --version
Version:      10.5.0
Git revision: 80b03db9
Git branch:   10-5-stable
GO version:   go1.8.5
Built:        2018-02-22T09:18:39+00:00
OS/Arch:      darwin/amd64
```

**Comment by niasharper** [comment]
Please let me know if you need any further details.  Thanks!

**Comment by ttaylorr** [comment]
Hi, thanks for opening such a detailed issue.

The error message that you posted about Git LFS skipping the local
checkout phase is when Git LFS detects that it is not installed (i.e.,
if you ran "git lfs pull" but hadn't run "git lfs install").

I think that what's going on here is that "git lfs install" when run
outside of a repository, will only install the LFS configuration in your
system (e.g., filter.lfs.process, etc.), but not the per-repo hooks.

You may need to run "git lfs install" in your repository, once per
repository, to ensure that things happen correctly.

The caveat is that some Git LFS tools will perform this step, but I
think that that depends on which you are running as a part of the
submodule checkout.

**Comment by niasharper** [comment]
@ttaylorr -- Thanks for the follow-up.  Two points of clarification here:

1 -- I believe that gitlab itself performs the git lfs install command upon initial clone
2 -- this particular case outlined above occurred on a workspace that was previously successfully cloned and successfully syncing.  After numerous builds in that workspace, a subsequent git fetch/git pull resulted in the "git lfs not installed" error state.

I'm digging for proof of point 1 in gitlab docs now.

To Re-iterate.... this error occurred on a repo that was already cloned and in a working state on previous executions.

**Comment by bk2204** [comment]
I think the situation that @niasharper is referring to is the GitLab CI system, which performs CI and builds much like Travis or AppVeyor does.

This message occurs because there isn't an entry for the `filter.lfs.clean` configuration in the repo in question.  Can you save the `.git/config` file when it's working and do a diff between the `.git/config` file when it's working and the one where it's broken and see if there's any other differences that might point to a relevant change?  Also, the timestamp of the `.git/config` file might point out when this issue occurs; you might be able to determine whether there's a correlation with a certain build step.

I'd like to see if there's a specific operation that git-lfs is doing that's causing this, or if not, try to help you determine what else is going on on your system that might cause this issue at that time.

**Comment by niasharper** [comment]
I've pulled a copy of .git/config and the .git/submodule/path/config -- also trying to track hooks for both the root repo and submodule repo.  Waiting for my problem state to re-occur now. I'll update when this comes up again.

Thanks!

**Comment by niasharper** [comment]
This just occurred again on a system in our CICD infrastructure.  It looks like the LFS failure is in the parent repo, not in the submodule repo (at least during this occurence).

The machine was previously pushing out builds just fine.  I performed a deep-dive into the .git configurations for both the root repo and the submodule repo.... the config file for both parent and submodule do not have any entries for filter.lfs.clean -- and the configurations for both submodules and parent modules between broken and working systems look the same.

Date on the config file between both parent and submodule shows an alteration at the same time the index was updated.  This might be the right track, if the thought is that some configurations are getting botched up.  It doesn't make much sense to me, though, since working systems have identical configurations.  The only differences between builders are unique tokens in the pathing and repo origin configuration that are specific to gitlab itself for authentication.

I think I may have found a possible root cause for this particular issue.  We have some custom code mucking with the git hooks on our root repo.  Our custom git hooks still include all of the LFS git hooks changes.... but the state of the parent repo (where the problem occurred) appears to be missing the post-checkout git hook entirely.

Git post-checkout from the LFS distribution should include:
```
# One time git hook setup
command -v git-lfs >/dev/null 2>&1 || { echo >&2 "\nThis repository is configured for Git LFS but 'git-lfs' was not found on your path. If you no longer wish to use Git LFS, remove this hook by deleting .git/hooks/post-checkout.\n"; exit 2; }
```
... We also include the following in the parent repo:
```
echo "*** cleaning submodules"
git submodule foreach --recursive git clean -fdx

echo "*** updating submodules"
git submodule update --init --recursive
```

The submodule post-checkout git hook exists properly as:
```
#!/bin/sh
command -v git-lfs >/dev/null 2>&1 || { echo >&2 "\nThis repository is configured for Git LFS but 'git-lfs' was not found on your path. If you no longer wish to use Git LFS, remove this hook by deleting .git/hooks/post-checkout.\n"; exit 2; }
git lfs post-checkout "$@"
```

I do not recall our parent repo post-checkout having the lfs post-checkout command before:
```
git lfs post-checkout "$@"
```
... Is this something from one of the newer LFS distributions?  What exactly is `$@` being expanded to and what would the impact be of excluding this command from the `./git/hooks/post-checkout` hook?

**Comment by niasharper** [comment]
nevermind on the ```git lfs post-checkout "$@"``` -- after a little more digging, this is a bash expansion for the passed command arguments.  Looks like our hook installer is silently failing and our resulting hook post-install is missing the git lfs post-checkout command.

I'm going to assume that this is our problem.  I'll re-open this ticket if I see anything else pop up with an LFS failure.

Thanks!

**Comment by niasharper** [comment]
We corrected our git hooks modifications.  We again ran into the error message in our logs "Skipping object checkout, Git LFS is not installed."

TL;DR -- I need to know exactly how the git lfs add-on identifies if it is installed in a repo??  ... and what exactly would `git lfs update --manual` update inside any repo metadata?  I went from not working to working in some processes that don't make sense.

Details below:

Here's the important excerpt from our gitlab build log:
```
$ ./common/config/git-hooks/install-git-hook.sh
git hooks installed successfully
$ git submodule foreach --recursive git clean -fdx
Entering '[path/to/submodule]'
$ git submodule update --init --recursive
$ git lfs pull
Skipping object checkout, Git LFS is not installed.```

What I had found previously is that the post-checkout git hook did not exist in .git/hooks due to the install script silently failing.  This is the entire install script being called:
```
#!/usr/bin/env bash
# One time git hook setup
ln -sf  ./../../common/config/git-hooks/post-checkout ./.git/hooks/post-checkout
RETVAL=$?
if [ $RETVAL != 0 ]; then
    echo Failed to symlink post-checkout with the return code $RETVAL
    exit 100
fi

ln -sf  ./../../common/config/git-hooks/post-merge ./.git/hooks/post-merge
RETVAL=$?
if [ $RETVAL != 0 ]; then
    echo Failed to symlink post-merge with the return code $RETVAL
    exit 101
fi

chmod a+x ./common/config/git-hooks/
chmod a+x ./common/config/git-hooks/post-checkout

if [ ! -L ./.git/hooks/post-checkout ]; then
    echo failed to confirm ./.git/hooks/post-checkout
    exit 200
fi

if [ ! -L ./.git/hooks/post-merge ]; then
    echo failed to confirm ./.git/hooks/post-merge
    exit 201
fi

echo git hooks installed successfully
exit 0
```

the post-checkout hook is exactly this:
```
#!/bin/bash

# One time git hook setup
command -v git-lfs >/dev/null 2>&1 || { echo >&2 "\nThis repository is configured for Git LFS but 'git-lfs' was not found on your path. If you no longer wish to use Git LFS, remove this hook by deleting .git/hooks/post-checkout.\n"; exit 2; }

echo "*** cleaning submodules"
git submodule foreach --recursive git clean -fdx

echo "*** updating submodules"
git submodule update --init --recursive

# git LFS post-checkout work
echo "*** performing LFS post-checkout"
git lfs post-checkout "$@"
```

I logged into the workspace and manually performed a `git lfs pull` and got the following to repro:
```
$ git lfs pull
Skipping object checkout, Git LFS is not installed.```

... okay... so I attempted a `git lfs install` as this usually corrects the issue.  Here's what came up:
```
$ git lfs install
Hook already exists: post-checkout

	#!/bin/bash
	
	# One time git hook setup
	command -v git-lfs >/dev/null 2>&1 || { echo >&2 "\nThis repository is configured for Git LFS but 'git-lfs' was not found on your path. If you no longer wish to use Git LFS, remove this hook by deleting .git/hooks/post-checkout.\n"; exit 2; }
	
	echo "*** cleaning submodules"
	git submodule foreach --recursive git clean -fdx
	
	echo "*** updating submodules"
	git submodule update --init --recursive
	
	# git LFS post-checkout work
	echo "*** performing LFS post-checkout"
	git lfs post-checkout "$@"

To resolve this, either:
  1: run `git lfs update --manual` for instructions on how to merge hooks.
  2: run `git lfs update --force` to overwrite your hook.
```
.... SO... I ran `git lfs update --manual` and got the following helper output:
```
$ git lfs  update --manual
Add the following to .git/hooks/pre-push:

	#!/bin/sh
	command -v git-lfs >/dev/null 2>&1 || { echo >&2 "\nThis repository is configured for Git LFS but 'git-lfs' was not found on your path. If you no longer wish to use Git LFS, remove this hook by deleting .git/hooks/pre-push.\n"; exit 2; }
	git lfs pre-push "$@"

Add the following to .git/hooks/post-checkout:

	#!/bin/sh
	command -v git-lfs >/dev/null 2>&1 || { echo >&2 "\nThis repository is configured for Git LFS but 'git-lfs' was not found on your path. If you no longer wish to use Git LFS, remove this hook by deleting .git/hooks/post-checkout.\n"; exit 2; }
	git lfs post-checkout "$@"

Add the following to .git/hooks/post-commit:

	#!/bin/sh
	command -v git-lfs >/dev/null 2>&1 || { echo >&2 "\nThis repository is configured for Git LFS but 'git-lfs' was not found on your path. If you no longer wish to use Git LFS, remove this hook by deleting .git/hooks/post-commit.\n"; exit 2; }
	git lfs post-commit "$@"

Add the following to .git/hooks/post-merge:

	#!/bin/sh
	command -v git-lfs >/dev/null 2>&1 || { echo >&2 "\nThis repository is configured for Git LFS but 'git-lfs' was not found on your path. If you no longer wish to use Git LFS, remove this hook by deleting .git/hooks/post-merge.\n"; exit 2; }
	git lfs post-merge "$@"```

... Hmm.... well... an ls command, and a cat command later -- yeah.. my hooks are still installed.  I wonder what would happen if I just did `git lfs pull` again... it should fail as I didn't update the hooks....
```$ git lfs pull
[goes back to prompt with success]```

Why does that work?

SO.... Back to my support question... exactly how does the git lfs add-on identifies if it is installed in a repo... and what exactly would `git lfs update --manual` update inside any repo metadata?

Is it as simple as looking for the exact below lines together in the hooks files?
```
	#!/bin/sh
	command -v git-lfs >/dev/null 2>&1 || { echo >&2 "\nThis repository is configured for Git LFS but 'git-lfs' was not found on your path. If you no longer wish to use Git LFS, remove this hook by deleting .git/hooks/post-checkout.\n"; exit 2; }
	git lfs post-checkout "$@"```

**Comment by bk2204** [comment]
The way we determine whether Git LFS is installed in this case is by looking for the `filter.lfs.clean` key within the Git configuration.  That could be within any configuration file valid for the repository: the repository-specific one, one belonging to the user, or the system-wide one.

I believe the reason we check for the filter specifically is because that’s what’s required to make the checkout work.

**Comment by niasharper** [comment]
That's interesting.... so, filter.lfs.clean should never be changing on these systems then if it's in a config --system or a config --global location.

I will verify again that the filter is in place.  I think I missed that this time around, unfortunately.  This was previously indicated in the thread.

For clarification, here's the local repo config output (super secret stuff redacted):
```
$ git config -l
core.excludesfile=~/.gitignore
core.legacyheaders=false
core.quotepath=false
mergetool.keepbackup=true
push.default=simple
color.ui=auto
color.interactive=auto
repack.usedeltabaseoffset=true
alias.s=status
alias.a=!git add . && git status
alias.au=!git add -u . && git status
alias.aa=!git add . && git add -u . && git status
alias.c=commit
alias.cm=commit -m
alias.ca=commit --amend
alias.ac=!git add . && git commit
alias.acm=!git add . && git commit -m
alias.l=log --graph --all --pretty=format:'%C(yellow)%h%C(cyan)%d%Creset %s %C(white)- %an, %ar%Creset'
alias.ll=log --stat --abbrev-commit
alias.lg=log --color --graph --pretty=format:'%C(bold white)%h%Creset -%C(bold green)%d%Creset %s %C(bold green)(%cr)%Creset %C(bold blue)<%an>%Creset' --abbrev-commit --date=relative
alias.llg=log --color --graph --pretty=format:'%C(bold white)%H %d%Creset%n%s%n%+b%C(bold blue)%an <%ae>%Creset %C(bold green)%cr (%ci)' --abbrev-commit
alias.d=diff
alias.master=checkout master
alias.spull=svn rebase
alias.spush=svn dcommit
alias.alias=!git config --list | grep 'alias\.' | sed 's/alias\.\([^=]*\)=\(.*\)/\1\     => \2/' | sort
include.path=~/.gitcinclude
include.path=.githubconfig
include.path=.gitcredential
diff.exif.textconv=exif
credential.helper=osxkeychain
filter.lfs.required=true
filter.lfs.clean=git-lfs clean -- %f
filter.lfs.smudge=git-lfs smudge -- %f
filter.lfs.process=git-lfs filter-process
fetch.recursesubmodules=false
http.[originURL].sslcainfo=**********************************
core.repositoryformatversion=0
core.filemode=true
core.bare=false
core.logallrefupdates=true
core.ignorecase=true
core.precomposeunicode=true
remote.origin.url=*************************************
remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
branch.master.remote=origin
branch.master.merge=refs/heads/master
submodule.[submodule].active=true
submodule.[submodule].url=**********************************************
```

Git System Config:
```
$ git config --system -l
core.excludesfile=~/.gitignore
core.legacyheaders=false
core.quotepath=false
mergetool.keepbackup=true
push.default=simple
color.ui=auto
color.interactive=auto
repack.usedeltabaseoffset=true
alias.s=status
alias.a=!git add . && git status
alias.au=!git add -u . && git status
alias.aa=!git add . && git add -u . && git status
alias.c=commit
alias.cm=commit -m
alias.ca=commit --amend
alias.ac=!git add . && git commit
alias.acm=!git add . && git commit -m
alias.l=log --graph --all --pretty=format:'%C(yellow)%h%C(cyan)%d%Creset %s %C(white)- %an, %ar%Creset'
alias.ll=log --stat --abbrev-commit
alias.lg=log --color --graph --pretty=format:'%C(bold white)%h%Creset -%C(bold green)%d%Creset %s %C(bold green)(%cr)%Creset %C(bold blue)<%an>%Creset' --abbrev-commit --date=relative
alias.llg=log --color --graph --pretty=format:'%C(bold white)%H %d%Creset%n%s%n%+b%C(bold blue)%an <%ae>%Creset %C(bold green)%cr (%ci)' --abbrev-commit
alias.d=diff
alias.master=checkout master
alias.spull=svn rebase
alias.spush=svn dcommit
alias.alias=!git config --list | grep 'alias\.' | sed 's/alias\.\([^=]*\)=\(.*\)/\1\     => \2/' | sort
include.path=~/.gitcinclude
include.path=.githubconfig
include.path=.gitcredential
diff.exif.textconv=exif
credential.helper=osxkeychain
```
... and our git global config:
```
$ git config --global -l
filter.lfs.required=true
filter.lfs.clean=git-lfs clean -- %f
filter.lfs.smudge=git-lfs smudge -- %f
filter.lfs.process=git-lfs filter-process
```

Unlike the git hooks, I'm fairly certain we're not doing any munging of the filters or other git config pieces.

**Comment by bk2204** [comment]
To answer the second question, `git lfs update --manual` doesn’t affect the filter configuration.  It fixes `lfs.*.access` if it’s wrong and prints the hook install steps, but otherwise doesn’t make any changes to the repository.

**Comment by bk2204** [comment]
Here’s a thought about what might be going on, which may or may not be off base: since the global configuration is where the LFS filter entries are, and this is per user, if `$HOME` is changed or the repository is accessed by a different user, then configuration wouldn’t be able to be accessed, and Git LFS would think it wasn’t installed.

I don’t know much about your configuration, so I can’t say whether that will be helpful to you.

**Comment by niasharper** [comment]
That's a good thought and catch.... unfortunately, it's not a different `$HOME` situation.  The gitlab runner process and all builds that have been failing this way all run under the same interactively logged in service account user.  I was logged in as this service account user when I pulled the above `git config` outputs as well.... so settings should be consistent across all of these occurrences.

**Comment by niasharper** [comment]
I think I found the problem... or at least the resulting error state.  

Just as you guys called it originally -- the filters are missing.  One of the other systems that started exhibiting this behavior appears to be missing filters in the local repo config and the global settings (build service account's `~/.gitconfig`) file is completely missing.

```
$ git config --global -l
fatal: unable to read config file '/Users/[build user]/.gitconfig': No such file or directory
```
Definitely not there:
```
$ cat ~/.gitconfig
cat: /Users/[build user]/.gitconfig: No such file or directory
```
... and I walked through all of my other boxes for the same user, `~/.gitconfig` does exist as expected and the output of `git config --global -l` outputs the expected lfs filters I previously provided.  Also -- all other local repos for this git lfs enabled project on the other build machines has the lfs filters in place in the local repo.

I'll be searching through all of my automation and customers' projects to see if anyone is doing bad things to the global configs.

Thanks!

**Comment by niasharper** [comment]
Just an update on this issue.... I still have not identified any scripts on my side which would be a cause resulting in the loss of the ~/.gitconfig file.  I had a number of failures with git lfs over the last few days.  It gave me confirmation of the error state... but no real source of the cause.

I took the opportunity to update from lfs 2.4.2 to 2.5.2 and will continue to monitor my build systems to see if the update makes an impact.

**Comment by giladbau** [comment]
Hi. I know this is an old thread, but I encountered this issue today and this thread helped me somewhat. However, I had to do some more digging to solve it on my end.  

I am using a Windows Server for my builds, and gitlab runner is using the SYSTEM user. Therefore, changing the global git config was not enough, since it does not affect other users. So I used 
`git config --system` to set my LFS variables, which solved the problem for all my repos.

Anyway, thought this should also be here if someone encounters this issue again.
